graph/pull.go

```diff
@@ -27,10 +18,28 @@ type ImagePullConfig struct {
 	OutStream   io.Writer
 }
 
+type Puller interface {
+	// Pull tries to pull `canonicalImageName` with tag `tag`, from `endpoint` with the config specified by `imagePullConfig`.
+	// `name` is the image's name on the server-side
+	// `sf` is an object that allows to stream back to the client (to be removed)
+	// Pull returns an error if any, as well as a boolean that determines whether to retry Pull on the next configured endpoint.
+	Pull(canonicalImageName, name, tag string, endpoint registry.APIEndpoint, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (fallback bool, err error)
```

Comment:
@stevvooe 

1. `name` is indeed `remoteName`, and instead of doing a bunch of cutting, I would prefer doing parsing and validation once with the `reference` package, and just pass the reference which should be able to provide all needed informations.
2. Yes, that's an obvious one, I did it already on my branch but then it induces a lot of trivial code changes that i would prefer to have in a separate PR for easier review.
3. `repoInfo` is going away, it's only there because of v1, and everything it provided, `reference` will provide it instead. Do you have any issues passing an `endpoint` to `Pull` ?

For the TODO I suggest the following:
```
/*
TODO(tiborvass):
    - use a `reference` package to remove ambiguity on the parameters, and remove the need for `repoInfo`.
    - put `sf` and `imagePullConfig` into the Puller creation method
```